python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1381 @@
+    """
+
+    cdef:
+        shared_ptr[CScanTask] wrapped
+        CScanTask* task
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__, subclasses_instead=False)
+
+    cdef init(self, shared_ptr[CScanTask]& sp):
+        self.wrapped = sp
+        self.task = self.wrapped.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CScanTask]& sp):
+        cdef SimpleScanTask self = SimpleScanTask.__new__(SimpleScanTask)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CScanTask] unwrap(self):
+        return self.wrapped
+
+    def scan(self):
+        """Iterate through sequence of materialized record batches.
+
+        Execution semantics are encapsulated in the particular ScanTask
+        implementation.
+
+        Returns
+        -------
+        record_batches : iterator of RecordBatch
+        """
+        cdef:
+            CRecordBatchIterator iterator
+            shared_ptr[CRecordBatch] record_batch
+
+        iterator = move(GetResultValue(move(self.task.Scan())))
+
+        while True:
+            iterator.Next(&record_batch)
+            if record_batch.get() == nullptr:
+                raise StopIteration()
+            else:
+                yield pyarrow_wrap_batch(record_batch)
+
+
+cdef class SimpleScanTask(ScanTask):
+    """A trivial ScanTask that yields the RecordBatch of an array."""
+
+    cdef:
+        CSimpleScanTask* simple_task
+
+    cdef init(self, shared_ptr[CScanTask]& sp):
+        ScanTask.init(self, sp)
+        self.simple_task = <CSimpleScanTask*> sp.get()
+
+
+cdef class ScannerBuilder:
+    """Factory class to construct a Scanner.
+
+    It is used to pass information, notably a potential filter expression and a
+    subset of columns to materialize.
+    """
+
+    cdef:
+        shared_ptr[CScannerBuilder] wrapped
+        CScannerBuilder* builder
+
+    def __init__(self, Dataset dataset not None, ScanContext context not None):
+        cdef shared_ptr[CScannerBuilder] builder
+        builder = make_shared[CScannerBuilder](
+            dataset.unwrap(), context.unwrap()
+        )
+        self.init(builder)
+
+    cdef void init(self, shared_ptr[CScannerBuilder]& sp):
+        self.wrapped = sp
+        self.builder = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CScannerBuilder]& sp):
+        cdef ScannerBuilder self = ScannerBuilder.__new__(ScannerBuilder)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CScannerBuilder] unwrap(self):
+        return self.wrapped
+
+    def project(self, columns):
+        """Set the subset of columns to materialize.
+
+        This subset will be passed down to DataSources and corresponding
+        DataFragments. The goal is to avoid loading, copying, and deserializing
+        columns that will not be required further down the compute chain.
+
+        Raises exception if any column name does not exists in the dataset's
+        Schema.
```

Comment:
Can you mention something about that this updates the ScannerBuilder in place (while still returning an object, so you can do method chaining I suppose?)

Reply:
Sure